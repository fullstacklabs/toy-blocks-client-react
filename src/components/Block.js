import React from "react";
import PropTypes from "prop-types";
import { Box, makeStyles, Typography } from "@material-ui/core";
import colors from "../constants/colors";

const Block = ({ block }) => {
  const classes = useStyles();
  return (
    <Box className={classes.box}>
      <div className={classes.block}>
        <Typography className={classes.index}>
          {String(block.attributes.index).padStart(3, '0')}
        </Typography>
        <Typography className={classes.data}>
          {block.attributes.data}
        </Typography>
      </div>
    </Box>
  );
};

const useStyles = makeStyles(() => ({
  box: {
    paddingBottom: '4px'
  },
  block: {
    background: 'rgba(0, 0, 0, 0.12)',
    paddingBottom: '4px',
    padding: '8px',
    font: 'Roboto'
  },
  index: {
    color: colors.blue,
    fontSize: '10px',
    fontWeight: 700,
    letterSpacing: '2px'
  },
  data: {
    fontSize: '14px',
    paddingBottom: '4px'
  }
}));

Block.propTypes = {
  block: PropTypes.shape({
    id: PropTypes.string,
    type: PropTypes.string,
    attributes: PropTypes.shape({
      index: PropTypes.number,
      timestamp: PropTypes.number,
      data: PropTypes.string,
      "previous-hash": PropTypes.string,
      hash: PropTypes.string
    }),
  }).isRequired,
};

export default Block;
