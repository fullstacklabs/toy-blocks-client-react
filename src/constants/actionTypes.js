export const CHECK_NODE_STATUS_START = 'CHECK_NODE_STATUS_START';
export const CHECK_NODE_STATUS_SUCCESS = 'CHECK_NODE_STATUS_SUCCESS';
export const CHECK_NODE_STATUS_FAILURE = 'CHECK_NODE_STATUS_FAILURE';

export const GET_BLOCK_NODE_STATUS_START = 'GET_BLOCK_NODE_STATUS_START';
export const GET_BLOCK_NODE_STATUS_FAILURE = 'GET_BLOCK_NODE_STATUS_FAILURE';
export const GET_BLOCK_NODE_STATUS_SUCCESS = 'GET_BLOCK_NODE_STATUS_SUCCESS';
