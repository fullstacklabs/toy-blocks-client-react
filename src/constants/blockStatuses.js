export const BLOCK_STATUSES = {
  IDDLE: 'iddle',
  LOADING: 'loading',
  FAILURE: 'failure',
  SUCCESS: 'success',
}
