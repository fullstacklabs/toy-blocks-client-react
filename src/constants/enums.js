export const LoadingStatus = {
  NOT_STARTED: "not started",
  LOADING: "loading",
  FINISHED: "finished",
  FAILURE: "failure",
};
