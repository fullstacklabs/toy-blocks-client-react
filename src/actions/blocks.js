import fetch from "cross-fetch";
import * as types from "../constants/actionTypes";

const fetchBlocksStatusStart = (node) => {
  return {
    type: types.FETCH_BLOCK_STATUS_START,
    node,
  };
};

const fetchBlocksStatusSuccess = (node, data) => {
  return {
    type: types.FETCH_BLOCK_STATUS_SUCCESS,
    node,
    data,
  };
};

const fetchBlocksStatusFailure = (node) => {
  return {
    type: types.FETCH_BLOCK_STATUS_FAILURE,
    node,
    data: null
  };
};

export function fetchBlocksList(node) {
  return async (dispatch) => {
    try {
      dispatch(fetchBlocksStatusStart(node));
      const res = await fetch(`${node}/api/v1/blocks`);

      if (res.status >= 400) {
        dispatch(fetchBlocksStatusFailure(node));
        return;
      }

      const json = await res.json();

      dispatch(fetchBlocksStatusSuccess(node, json?.data));
    } catch (err) {
      dispatch(fetchBlocksStatusFailure(node));
    }
  };
}

export function fetchBlocksLists(node) {
  return (dispatch) => {
    dispatch(fetchBlocksList(node))
  }
}
