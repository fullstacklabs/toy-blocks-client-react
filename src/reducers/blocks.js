import {
  FETCH_BLOCK_STATUS_START,
  FETCH_BLOCK_STATUS_SUCCESS,
  FETCH_BLOCK_STATUS_FAILURE
} from '../constants/actionTypes';

import { BLOCK_STATUSES } from '../constants/blockStatuses';

import initialState from './initialState';

export default function nodesReducer(state = initialState().blocks, action) {
  const storedNode = state?.blocks?.[action?.node] ?? {};
  switch (action.type) {
    case FETCH_BLOCK_STATUS_START:
      return {
        ...state,
        [action.node]: {
          ...storedNode,
          status: BLOCK_STATUSES.LOADING,
        }
      };
    case FETCH_BLOCK_STATUS_SUCCESS:
      return {
        ...state,
        [action.node]: {
          ...storedNode,
          status: BLOCK_STATUSES.SUCCESS,
          data: action.data,
        }
      };
    case FETCH_BLOCK_STATUS_FAILURE:
      return {
        ...state,
        [action.node]: {
          status: BLOCK_STATUSES.FAILURE,
          data: action.data ?? null,
        }
      };
    default:
      return state;
  }
}
