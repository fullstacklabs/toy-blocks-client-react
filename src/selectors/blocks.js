import { BLOCK_STATUSES } from '../constants/blockStatuses';

export const getNodesBlocks = (state) => state.blocks ?? {};

export const getNodeBlocks = (state, node) => {
  const blocks = getNodesBlocks(state);
  return blocks[node?.url] ?? {};
}

export const blockIsLoading = (state, node) => {
  const block = getNodeBlocks(state, node);
  return block?.status === BLOCK_STATUSES.LOADING
}

export const blockHadFailure = (state, node) => {
  const block = getNodeBlocks(state, node);
  return block?.status === BLOCK_STATUSES.FAILURE
}

export const blockIsIddle = (state, node) => {
  const block = getNodeBlocks(state, node);
  return block?.status === BLOCK_STATUSES.IDDLE
}

