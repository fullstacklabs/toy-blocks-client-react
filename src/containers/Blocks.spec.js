import React from "react";
import { shallow } from "enzyme";
import configureMockStore from "redux-mock-store";
import thunk from 'redux-thunk';
import { Provider } from "react-redux";
import { create } from "react-test-renderer";
import ConnectedBlocks, { Blocks } from "./Blocks";
import Block from "../components/Block";

import {
  CircularProgress,
} from "@material-ui/core";


import initialState from '../reducers/initialState';

import { BLOCK_STATUSES } from '../constants/blockStatuses';
import { getNodeBlocks, blockIsLoading, blockHadFailure, blockIsIddle } from "../selectors/blocks";


const node = {
  url: 'http://localhost',
  online: true,
  name: 'Node 1',
  loading: false
};

const fakeBlock = {
  attributes: {
    data: "Lorem ipsum data",
    hash: "aaaaaaaaaa",
    index: 1,
  },
  id: 1,
  type: "blocks"
}

describe("<Blocks />", () => {

  const getInitialState = () => {
    return initialState().blocks;
  };

  it("Block is in Iddle", () => {
    const newState = getInitialState();
    const props = {
      retryRequest: jest.fn(),
      blocks: getNodeBlocks(newState, node),
      isLoading: blockIsLoading(newState, node),
      isFailured: blockHadFailure(newState, node),
      isIddle: blockIsIddle(newState, node),
    }
    const wrapper = shallow(
      <Blocks node={node} {...props} />
    );
    expect(wrapper.find(Block).length).toEqual(0);
  });

  it("Block is loading", () => {
    const newState = {...getInitialState(), [node.url]: {
      status: BLOCK_STATUSES.LOADING
    }}
    const props = {
      retryRequest: jest.fn(),
      blocks: getNodeBlocks(newState, node),
      isLoading: blockIsLoading(newState, node),
      isFailured: blockHadFailure(newState, node),
      isIddle: blockIsIddle(newState, node),
    }

    const wrapper = shallow(
      <Blocks node={node} {...props} />
    );
    expect(wrapper.find(CircularProgress).length).toEqual(0);
  });

  it("Block had failured", () => {
    const newState = {...getInitialState(), [node.url]: {
      status: BLOCK_STATUSES.FAILURE
    }}
    const props = {
      retryRequest: jest.fn(),
      blocks: getNodeBlocks(newState, node),
      isLoading: blockIsLoading(newState, node),
      isFailured: blockHadFailure(newState, node),
      isIddle: blockIsIddle(newState, node),
    }

    const wrapper = shallow(
      <Blocks node={node} {...props} />
    );
    expect(wrapper.find(CircularProgress).length).toEqual(0);
  });

  it("Block request was successfull", () => {
    const newState = {...getInitialState(), [node.url]: {
      status: BLOCK_STATUSES.SUCCESS,
      data: new Array(3).fill(fakeBlock)
    }}

    const props = {
      retryRequest: jest.fn(),
      blocks: getNodeBlocks(newState, node),
      isLoading: blockIsLoading(newState, node),
      isFailured: blockHadFailure(newState, node),
      isIddle: blockIsIddle(newState, node),
    }

    const wrapper = shallow(
      <Blocks node={node} {...props} />
    );

    expect(wrapper.find(Block).length).toBeGreaterThanOrEqual(0);
  });

  it("should match snapshot", () => {
    const middlewares = [thunk];
    const store = configureMockStore(middlewares)(getInitialState());
    const component = create(
      <Provider store={store}>
        <ConnectedBlocks />
      </Provider>
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
