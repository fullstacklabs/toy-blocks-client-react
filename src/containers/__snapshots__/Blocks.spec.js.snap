// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Blocks /> should match snapshot 1`] = `
<div
  className="MuiBox-root MuiBox-root-8"
/>
`;
